docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+        model 2          :a22, after a21, until a32
+
+        Trigger3 :milestone, t3, 1945, 0d
+        train_batch_3 :a30, after a20, until t3
+        training3 :a31, after t3, 1y
+        model 3          :a32, after a31, until a42
+
+        Trigger4 :milestone, t4, 1950, 0d
+        train_batch_4 :a40, after a30, until t4
+        training4 :a41, after t4, 1y
+        model 4          :a42, after a41, 10y
+
+```
+
+With our assumption of `training_time=0` this `training_<i>` would be can be removed so that we can evaluate one model on the data between two triggers (hence the next training batch)
+
+##### State of modyn (at the time of writing)
+
+We currectly evaluate a model directly after training it (e.g. we are now at `Trigger 4`) and have now robust approach of combining the results of the single model evaluations to a pipeline performance. We support a `OffsetEvaluationStrategy` that specifies the interval in which we want to evaluate the models. We also support `MatrixEvaluationStrategy`: We evaluate on predefined grid of evaluation points.
+
+We face several problems:
+
+1) We don't know when the next trigger occurs.
+    This might be fine if we assume a time based trigger policy with a regular schedule where we can set the OffsetEvaluationStrategy lookahead offset to the trigger interval. However, for a generic trigger policy we cannot assume this. In consequence when setting a fixed lookahead offset we might end up evaluating a model with a lookahead offset that is too large or too small. (E.g. when always evaluate on the next 5 years after a training. That would mean that if a policy doesn't retrain for 20 years we miss a lot of data)
+
+> After `Trigger 4` we e.g. could end up only considering the first 5 years of the time where `model 4` is used. That could completely skew the pipeline performance if e.g. as heavy drift occurs after 5 years of using `model 4`.
+
+2) As mentioned above the individual performances of every evaluation cannot simple be averaged to get the pipeline performance as number of samples in every time range is important!
+
+3) Without making any assumptions on how & when to trigger `MatrixEvaluationStrategy` is not of any use as the grid splits are arbitrary and don't correspond to the times when triggers occur and models are transitioned.
+
+4) `OffsetEvaluationStrategy` is not flexible enough to handle arbitrary intervals. Even if we would run after the pipeline where we know the exact points in time when triggers occurred, we could still not specify the intervals as `OffsetEvaluationStrategy` only supports 1-sided intervals. It currently assumes being given `first_timestamp` and `last_timestamp` representing the first and last sample timestamp for the data batch that lead to the trigger. These do not correspond to the start and end time of the model usage.
+
+>>> The ideal solution would therefore evaluate models on their entire usage period with all samples in that period.
```

Comment:
I think that is _one_ evaluation policy, but I might be wrong. I don't think we should remove e.g. the evaluation on the matrix for the reasons mentioned before. Yes, the performance of a model while it is being used/most recent is probably the most interesting metric, but for experiments, evaluating the model from year 2006 on data from 2010 might still be interesting.

Reply:
Agreed, let's keep the matrix evaluation!

Reply:
Yes, for sure we can extend it - as long as we somehow are able to express the original idea, that's fine :D